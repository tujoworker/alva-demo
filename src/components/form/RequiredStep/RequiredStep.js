/**
 * Component
 *
 * @author Tobias HÃ¸egh <tobias@tujo.no>
 */

// import './RequiredStep.css'
// import '../../../styles/form.css'

import { observer, inject } from 'mobx-react'
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import anime from 'animejs'

import { FormattedMessage } from 'react-intl'

import Config from '../../../config/Config'
import Container from '../../Wrapper'
import ContinueCardQuestion from '../Dialogs/ContinueCardQuestion'
import FormModel from './RequiredStep.schema'
import FormStyle from '../FormStyles'
import SimpleInput from '../../form/inputs/SimpleInput'

@withRouter
@inject('store')
@observer
export default class Form extends Component {
  static propTypes = {
    store: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired,
    onSubmit: PropTypes.oneOf([PropTypes.function, null])
  }
  static defaultProps = {
    onSubmit: null
  }
  state = {}
  constructor(props) {
    super(props)
    this.form = new FormModel({
      // onSuccess: () => this.submit(),
      // onError: form => {
      // 	console.log('All form errors', form.errors())
      // }
    })
    // console.log('this.form', this.form)
    this.form.$hooks.onSuccess = form => {
      console.log('Form Values!', form.values())
      this.submit()
    }
    this.form.$hooks.onError = form => {
      console.log('All form errors', form.errors())
    }
  }
  componentWillMount() {
    this.form_disposer = this.props.store.firstStep.bindForm(this.form)
    // this.form_disposer = this.props.store.regForm.bindForm(this.form)
  }
  componentDidMount() {
    anime(
      Object.merge(Config.viewTransactionAnimationIn, {
        targets: '.app__root form',
        delay: 100
      })
    )
  }
  componentWillUnmount() {
    this.form_disposer()
  }
  // onSuccess() {
  // 	console.log('onSuccess')
  // 	this.props.history.push('/form/second')
  // }
  applyNewRegData() {
    this.props.store.regForm.applyNewRegData()
    if (!this.props.store.regForm.props.navigation.location)
      this.props.history.push('/form/first')
  }
  submit() {
    if (this.props.onSubmit) return this.props.onSubmit()
    anime(
      Object.merge(Config.viewTransactionAnimationOut, {
        targets: '.app__root form',
        complete: () => {
          this.props.history.push('/form/first')
          // this.props.store.regForm.setLocation('/form/first')
        }
      })
    )
  }
  render() {
    // const { componentState } = this.props.store.firstStep
    return (
      <Container>
        <div css={FormStyle}>
          <h1>Name: {this.props.store.regForm.props.fields.name}</h1>
          {/* <h2>nr: {this.props.store.firstStep.props.fields.nr}</h2> */}
          {this.props.store.firstStep.showContinueCardQuestion && (
            <ContinueCardQuestion
              onSubmit={this.applyNewRegData.bind(this)}
            />
          )}
          <form onSubmit={this.form.onSubmit}>
            <fieldset>
              <SimpleInput field={this.form.$('nr')} />

              <br />
              <button
                type="submit"
                className="action-button"
                onClick={this.form.onSubmit}
              >
                <FormattedMessage id="form.next.label" />
              </button>

              <p>{this.form.error}</p>
            </fieldset>
          </form>
        </div>
      </Container>
    )
  }
}
