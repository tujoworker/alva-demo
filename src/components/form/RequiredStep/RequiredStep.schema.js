/**
 * Form Model
 *
 * @author Tobias HÃ¸egh <tobias@tujo.no>
 */

import { Form } from 'mobx-react-form' // https://foxhound87.github.io/mobx-react-form/docs
import validatorjs from 'validatorjs' // https://github.com/skaterdav85/validatorjs
import { L } from '../../../core/startup/locales'

// import history from '../../../core/routing/History'

export default class FormModel extends Form {
  constructor(props) {
    super(props)
    this.props = props
  }

  setup() {
    return {
      fields: [
        {
          name: 'nr',
          label: L('form.nr.label'),
          placeholder: L('form.nr.placeholder'),
          rules: ['required', 'numeric', 'regex:/^[0-9]{6,11}$/']
        }
      ]
    }
  }

  hooks() {
    return {
      // onSuccess(form) {
      // 	this.props.onSuccess(form.values())
      // },

      onError(form) {
        console.log('All form errors', form.errors())
        // this.props.onError(form.values())
      }
    }
  }

  plugins() {
    return { dvr: validatorjs } //dvr = Declarative Validation Rules / validatorjs
  }
}
