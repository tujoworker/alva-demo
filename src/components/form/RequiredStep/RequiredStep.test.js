/**
 * A simple test, to test the results of components
 *
 * @author Tobias HÃ¸egh <tobias@tujo.no>
 */

//test core packages
// import { shallow, mount, fakeProps } from 'alva'
// import { getRouteProps } from 'alva/static'
import { shallow, mount, fakeProps } from 'alva/tdd'
import React from 'react'

import toJson from 'enzyme-to-json'

import RequiredStep from './RequiredStep'
import ProviderProps from '../../../core/test/ProviderProps'

// The API Docs
// http://airbnb.io/enzyme/docs/api/
// https://facebook.github.io/jest/docs/en/expect.html

// some handy utilities
// learn more about this `sel` function
// from my other blog post: http://kcd.im/sel-util
// const sel = id => `[data-test="${id}"]`
// const hasMenu = wrapper => wrapper.find(sel('menu')).length === 1
//

describe('Component renders correctly', () => {
  // only locale={localeStore} is needed from the Provider
  const HeaderWrap = mount(
    <ProviderProps.Provider lang="en">
      <RequiredStep.wrappedComponent {...ProviderProps} />
    </ProviderProps.Provider>
  )

  // console.log('toJson(HeaderWrap)', toJson(HeaderWrap).find('form'))

  // Once we have the Provider in the three, the toJson gets way to large!
  it('have to match snapshot', () => {
    expect(toJson(HeaderWrap.find('Container'))).toMatchSnapshot()
  })

  it('has a form tag', () => {
    expect(HeaderWrap.find('form').exists()).toBe(true)
    // expect(HeaderW rap.first().props().role).toBe('banner')
  })
  it('has a fieldset tag', () => {
    expect(HeaderWrap.find('fieldset').exists()).toBe(true)
  })
  it('has a submit button', () => {
    expect(HeaderWrap.find('button[type="submit"]').exists()).toBe(true)
  })
  it('has ... after sumit with empty fields', () => {
    HeaderWrap.find('button[type="submit"]').simulate('click')
    expect(HeaderWrap.find('span.err_msg').exists()).toBe(true)
  })
  it('has a FormattedMessage with the content of "Next"', () => {
    expect(HeaderWrap.find('FormattedMessage').text()).toMatch(/Next/)
  })

  // it('has a menu inside', () => {
  //   expect(HeaderWrap.closest('nav').exists()).toBe(true)
  // })

  // it('has to have at least one li element', () => {
  //     expect(HeaderWrap.find('li.list-group-item')).toHaveLength(1);
  // });
})
