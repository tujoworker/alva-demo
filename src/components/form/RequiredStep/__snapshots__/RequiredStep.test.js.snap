// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders correctly have to match snapshot 1`] = `
<Container>
  <div
    className="app__content"
    css="css-1uifu4x"
  >
    <div
      css="css-10zrkg5"
    >
      <div
        css="css-1eb3d0b"
      >
        <h1>
          Name: 
        </h1>
        <form
          onSubmit={[Function]}
        >
          <fieldset>
            <Input
              field={
                Field {
                  "$bindings": "default",
                  "$changed": false,
                  "$clearing": false,
                  "$default": "",
                  "$disabled": false,
                  "$extra": null,
                  "$focused": false,
                  "$handlers": Object {},
                  "$hooks": Object {},
                  "$initial": "",
                  "$input": [Function],
                  "$interceptors": null,
                  "$label": "Nr.",
                  "$observers": null,
                  "$options": Object {},
                  "$output": [Function],
                  "$placeholder": "Your Nr.",
                  "$related": Array [],
                  "$resetting": false,
                  "$rules": Array [
                    "required",
                    "numeric",
                    "regex:/^[0-9]{6,11}$/",
                  ],
                  "$submitting": false,
                  "$touched": false,
                  "$validatedWith": "value",
                  "$validating": false,
                  "$validators": null,
                  "$value": "",
                  "autoFocus": false,
                  "debouncedValidation": [Function],
                  "disposeValidationOnBlur": [Function],
                  "errorAsync": null,
                  "errorSync": "The Nr. field is required.",
                  "execHandler": [Function],
                  "execHook": [Function],
                  "fields": Object {},
                  "files": undefined,
                  "hasInitialNestedFields": false,
                  "id": "nr--1",
                  "incremental": false,
                  "intercept": [Function],
                  "key": "nr",
                  "name": "nr",
                  "noop": [Function],
                  "observe": [Function],
                  "onAdd": [Function],
                  "onBlur": [Function],
                  "onChange": [Function],
                  "onClear": [Function],
                  "onDel": [Function],
                  "onDrop": [Function],
                  "onFocus": [Function],
                  "onReset": [Function],
                  "onSubmit": [Function],
                  "onToggle": [Function],
                  "path": "nr",
                  "showError": false,
                  "state": State {
                    "$struct": Array [],
                    "bindings": Bindings {
                      "rewriters": Object {
                        "default": Object {
                          "autoFocus": "autoFocus",
                          "checked": "checked",
                          "disabled": "disabled",
                          "id": "id",
                          "label": "label",
                          "name": "name",
                          "onBlur": "onBlur",
                          "onChange": "onChange",
                          "onFocus": "onFocus",
                          "placeholder": "placeholder",
                          "type": "type",
                          "value": "value",
                        },
                      },
                      "templates": Object {},
                    },
                    "current": Object {
                      "fields": Object {},
                      "props": Object {},
                    },
                    "disposers": Object {
                      "interceptor": Object {},
                      "observer": Object {},
                    },
                    "form": FormModel {
                      "$clearing": false,
                      "$handlers": Object {},
                      "$hooks": Object {
                        "onError": [Function],
                        "onSuccess": [Function],
                      },
                      "$resetting": false,
                      "$submitting": false,
                      "$validating": false,
                      "debouncedValidation": [Function],
                      "execHandler": [Function],
                      "execHook": [Function],
                      "fields": Object {
                        "nr": [Circular],
                      },
                      "intercept": [Function],
                      "name": null,
                      "noop": [Function],
                      "observe": [Function],
                      "onAdd": [Function],
                      "onClear": [Function],
                      "onDel": [Function],
                      "onReset": [Function],
                      "onSubmit": [Function],
                      "props": Object {
                        "fields": Array [
                          Object {
                            "label": "Nr.",
                            "name": "nr",
                            "placeholder": "Your Nr.",
                            "rules": Array [
                              "required",
                              "numeric",
                              "regex:/^[0-9]{6,11}$/",
                            ],
                          },
                        ],
                      },
                      "state": [Circular],
                      "validator": Validator {
                        "drivers": Object {
                          "dvr": DVR {
                            "asyncRules": Array [],
                            "extend": undefined,
                            "options": Options {
                              "options": Object {
                                "allowRequired": false,
                                "autoParseNumbers": false,
                                "defaultGenericError": null,
                                "retrieveOnlyDirtyValues": false,
                                "retrieveOnlyEnabledFields": false,
                                "showErrorsOnBlur": true,
                                "showErrorsOnChange": true,
                                "showErrorsOnClear": false,
                                "showErrorsOnInit": false,
                                "showErrorsOnReset": true,
                                "showErrorsOnSubmit": true,
                                "strictDelete": true,
                                "strictUpdate": false,
                                "submitThrowsError": true,
                                "uniqueId": [Function],
                                "validateDisabledFields": false,
                                "validateOnBlur": true,
                                "validateOnChange": false,
                                "validateOnInit": true,
                                "validationDebounceOptions": Object {
                                  "leading": false,
                                  "trailing": true,
                                },
                                "validationDebounceWait": 250,
                              },
                            },
                            "promises": Array [],
                            "validator": [Function],
                            "validators": Object {},
                          },
                          "vjf": VJF {
                            "options": Options {
                              "options": Object {
                                "allowRequired": false,
                                "autoParseNumbers": false,
                                "defaultGenericError": null,
                                "retrieveOnlyDirtyValues": false,
                                "retrieveOnlyEnabledFields": false,
                                "showErrorsOnBlur": true,
                                "showErrorsOnChange": true,
                                "showErrorsOnClear": false,
                                "showErrorsOnInit": false,
                                "showErrorsOnReset": true,
                                "showErrorsOnSubmit": true,
                                "strictDelete": true,
                                "strictUpdate": false,
                                "submitThrowsError": true,
                                "uniqueId": [Function],
                                "validateDisabledFields": false,
                                "validateOnBlur": true,
                                "validateOnChange": false,
                                "validateOnInit": true,
                                "validationDebounceOptions": Object {
                                  "leading": false,
                                  "trailing": true,
                                },
                                "validationDebounceWait": 250,
                              },
                            },
                            "promises": Array [],
                            "validator": null,
                          },
                        },
                        "error": null,
                        "form": [Circular],
                        "options": Object {},
                        "plugins": Object {
                          "dvr": [Function],
                          "svk": false,
                          "vjf": true,
                        },
                        "promises": Array [],
                        "schema": Object {},
                      },
                    },
                    "initial": Object {
                      "fields": Object {},
                      "props": Object {},
                    },
                    "mode": "unified",
                    "options": Options {
                      "options": Object {
                        "allowRequired": false,
                        "autoParseNumbers": false,
                        "defaultGenericError": null,
                        "retrieveOnlyDirtyValues": false,
                        "retrieveOnlyEnabledFields": false,
                        "showErrorsOnBlur": true,
                        "showErrorsOnChange": true,
                        "showErrorsOnClear": false,
                        "showErrorsOnInit": false,
                        "showErrorsOnReset": true,
                        "showErrorsOnSubmit": true,
                        "strictDelete": true,
                        "strictUpdate": false,
                        "submitThrowsError": true,
                        "uniqueId": [Function],
                        "validateDisabledFields": false,
                        "validateOnBlur": true,
                        "validateOnChange": false,
                        "validateOnInit": true,
                        "validationDebounceOptions": Object {
                          "leading": false,
                          "trailing": true,
                        },
                        "validationDebounceWait": 250,
                      },
                    },
                    "strict": false,
                  },
                  "sync": [Function],
                  "type": "text",
                  "validationAsyncData": Object {},
                  "validationErrorStack": Array [
                    "The Nr. field is required.",
                  ],
                  "validationFunctionsData": Array [],
                }
              }
              placeholder={null}
              type={null}
            >
              <div
                className="input_set"
              >
                <label
                  htmlFor="nr--1"
                >
                  Nr.
                </label>
                <span
                  className="form_con"
                >
                  <input
                    autoFocus={false}
                    checked={null}
                    disabled={false}
                    id="nr--1"
                    label="Nr."
                    name="nr"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Your Nr."
                    type="text"
                    value=""
                  />
                </span>
              </div>
            </Input>
            <br />
            <button
              className="action-button"
              onClick={[Function]}
              type="submit"
            >
              <FormattedMessage
                id="form.next.label"
                values={Object {}}
              >
                <span>
                  Next
                </span>
              </FormattedMessage>
            </button>
            <p />
          </fieldset>
        </form>
      </div>
    </div>
  </div>
</Container>
`;
